Add Shootout Cog

ShroomAgent27 opened the pull request:
* Implements Suggestion #394 on Champions of Equestria's Discord Server
* Rewards players for fast typing
* Configurable message to type to win.
* Settable pot amount, even zero.
* Configurable wait time after the first player has joined the game.

ShroomAgent27 committed:
Add Shootout Cog

* Implements Suggestion #394 on Champions of Equestria's Discord Server
* Rewards players for fast typing
* Configurable message to type to win.
* Settable pot amount, even zero.
* Configurable wait time after the first player has joined the game.

brandons209 commented on shootout/shootout.py:
```diff
@@ -0,0 +1,173 @@
+
+    @commands.guild_only()
+    @checks.admin_or_permissions(administrator=True)
+    @commands.command(hidden=True)
```
I wouldn't make this hidden, should be seen by admins who run help command.

ShroomAgent27 replied:
Debug config dump and reset commands are now visible.

brandons209 commented on shootout/shootout.py:
```diff
@@ -0,0 +1,173 @@
+            return
+        listen_message = await self.config.guild(ctx.guild).message()
+        await self.config.guild(ctx.guild).Session.Active.set(True)
+        await ctx.send(
```
So, a few things. This should send out this message, but instead of people just typing the message, it should say something like: "When I say draw! type \<msg\> to shoot!" and then randomize a time to wait in some range, then say "Draw!" and thats when the players type the \<msg\>. The issue with the way it is now is that it always waits the same amount of time and becomes predictable.

ShroomAgent27 replied:
Added delay and fuzzy config values to add a somewhat randomized wait time between when the lobby ends and the game begins.

brandons209 commented on shootout/shootout.py:
```diff
@@ -0,0 +1,173 @@
+            await ctx.send("A shootout with yourself means just shooting at some bottles. For that, no charge.")
+            await self.config.guild(ctx.guild).Session.clear()
+            return
+        listen_message = await self.config.guild(ctx.guild).message()
```
This listen message shouldn't be just one message, it should be a list of configurable messages so its not always the same. here it should randomly select one from the list.

ShroomAgent27 replied:
Added configurable listen/draw messages.

brandons209 commented on shootout/shootout.py:
```diff
@@ -0,0 +1,173 @@
+        except BalanceTooHigh as e:
+            await bank.set_balance(winner, e.max_balance)
+        await ctx.send(
+            "{} has won the shootout! {} {} has been deposited to their account!".format(
```
Some randomized win messages would be cool, not a big deal though. Like Russian Roulette

brandons209 commented on shootout/shootout.py:
```diff
@@ -0,0 +1,173 @@
+        return
+
+    @soset.command(name="wait")
+    async def soset_wait(self, ctx, wait: int):
```
I would set an upper limit, because too long can cause issues.

ShroomAgent27 replied:
Added warnings when setting excessively high wait times.

ShroomAgent27 committed:
Implement requested changes.

* Made a delay between when the shootout starts and when messages can be submitted.
    - This delay can be configurable.
* Added a fuzzy time to help randomize the time between when the lobby ends and shootout begins.
    - This fuzzy time can be configurable.
* "Draw!" messages are now configurable. This is the message the bot listens for.
* Added randomized and configurable victory messages.
    - The first {} is always the user's mention.
    - The second {} is always the currency amount won.
    - The final {} is always the name of the currency.
* Added warnings when setting excessively high wait times. Admins can confirm they want high wait times with a "yes" message.
* Debug config dump and reset commands are now visible.

brandons209 commented on shootout/shootout.py:
```diff
@@ -3,28 +3,27 @@
-        "message": "Shoot!",
-        "wait_time": 60,
+        "Messages": ["Shoot!", "Draw!", "Fire!", "Kill!"],
         "Session": {"Pot": 0, "Players": [], "Active": False},
```
So your code to play games works in a single channel, but the way Session is setup means only one game at a time can be played per guild. You should set session to the channel config, so every text channel can have an independent game going. Also need to then change the admin clear_game function to clear game for the channel

ShroomAgent27 replied:
Implemented multi-channel support.

brandons209 commented on shootout/shootout.py:
```diff
@@ -142,32 +152,132 @@ async def soset_bet(self, ctx, amount: int):
-        """Sets the message that the bot listens for during shootout games. Only affects new games!"""
-        await self.config.guild(ctx.guild).message.set(message)
+    @soset.command(name="draw")
+    async def soset_draw(self, ctx, *, draws: str):
```
If no draws are given as an argument, this should list all draws currently set. You can do this by adding a default value of None to draws argument and checking if draws is None in function.

ShroomAgent27 replied:
Implemented.

brandons209 commented on shootout/shootout.py:
```diff
@@ -142,32 +152,132 @@ async def soset_bet(self, ctx, amount: int):
 
-    @soset.command(name="wait")
+    @soset.command(name="victory")
+    async def soset_victory(self, ctx, *, victories: str):
```
Few things. First, reference the comment above for the soset_draw function about listing currently set victories.

The help text for this function doesn't mention you have to include {} brackets in your victory messages so the proper stuff gets set. And as you mention here will cause errors, which isn't good and not user friendly. What you should do is look at the welcome cog on how it conditionally formats messages. You should see use named parameters for format method of strings. 

Then, in your help text here you can mention: `Use {winner} to specify the winner, {amount} to specify the amount` etc. In your code to format the victory message you would do something like `VICTORY_MSG.format(winner=m.author.mention, amount=pot)`. 

https://www.dummies.com/programming/python/use-named-arguments-in-format-strings/

This way, no errors will happen if people forget to add the {winner} or other things in their victory messages.
Maybe even put in a check if you get victory messages with one word to give them a warning.

Finally, a suggestion I have to improving the ease of use is to make separate functions to add/delete victories, same with draw messages above. If you don't do this, then if I want to add a single victory message, I would first have to copy all the previous ones, then add mine, then run the command, which isn't very user friendly. I know its easier to do it this way in code but ultimately it'll be better off setting it up like that.

ShroomAgent27 replied:
* Implemented victory add, victory list, and victory rem command that functions similar to paths in CogManagerUI.
* Implemented user friendly argument format in victory messages.
* Adds a warning if not all arguments are present in the victory message when adding.

brandons209 commented on shootout/shootout.py:
```diff
@@ -100,13 +103,20 @@ async def end_game(self, ctx, winner: discord.Member):
-                winner.mention, total, currency
-            )
-        )
+        await ctx.send((await self.get_random_win_message(ctx)).format(winner.mention, total, currency))
```
See the victory message comment down below.

brandons209 reviewed (changes requested):
Great code, a few more touch ups and this will be a fun little cog!

ShroomAgent27 committed:
Implement multi-channel support.

* Can now have shootouts in multiple channels.

ShroomAgent27 committed:
Display current draws list if none is provided.

ShroomAgent27 committed:
Make inputting victory messages more user friendly.

* Have a list command to post all current victories.
* Have an add command to be able to add a victory without listing the others.
* Have a remove command to be able to remove one victory without removing others.

ShroomAgent27 committed:
Use user friendly conditional victory formatting.

* Implement {winner}, {amount}, and {currency} fields for formatting.
* Prints a warning if a tag isn't in the victory message.
    - One warning is printed per tag missing.

ShroomAgent27 committed:
Remove debug markers.

brandons209 commented on shootout/shootout.py:
```diff
@@ -0,0 +1,385 @@
+        msg = "Victories:\n\n"
+        partial = []
+        for index, victory in enumerate(victories, start=1):
+            partial.append("{}. {}".format(index, victory))
```
You should put everything in a box using the box() chat formatting. You should also pagify in case discord message limit is reached.

shroomdog27 replied:
Changed to 

https://github.com/shroomdog27/brandons-cogs/blob/fb5d2d6900497777a79ac3cd7e04741a12d9630e/shootout/shootout.py#L266-L268

brandons209 commented on shootout/shootout.py:
```diff
@@ -0,0 +1,385 @@
+        return
+
+    @soset.command(name="draw")
+    async def soset_draw(self, ctx, *, draws: str = None):
```
Since victories has add/remove/list commands, you should do the same thing for draws.

ShroomAgent27 committed:
Pagify and box messages for Victory list.

ShroomAgent27 committed:
Implement list/add/rem commands for draw messages.

brandons209 reviewed (approved):
its lit now. good job.
